src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,46 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private int size;
+    private Node<T> first;
 
+    public void push(T value) {
+        first = new Node<>(value, first);
+        size++;
     }
 
     public T peek() {
-        return null;
+        emptyStack();
+        return first.value;
     }
 
     public T pop() {
-        return null;
+        emptyStack();
+        T value = first.value;
+        first = first.next;
+        size--;
+        return value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private void emptyStack() {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node next;
+
+        public Node(T value, Node<T> next) {
```

Comment:
is the constructor available from outside if the Node class is private?
```suggestion
        private Node(T value, Node<T> next) {
```